src/board/data-logging.ts

```diff
@@ -0,0 +1,225 @@
+import { LogEntry } from ".";
+import {
+  MICROBIT_HAL_DEVICE_NO_RESOURCES,
+  MICROBIT_HAL_DEVICE_OK,
+  MICROBIT_HAL_LOG_TIMESTAMP_DAYS,
+  MICROBIT_HAL_LOG_TIMESTAMP_HOURS,
+  MICROBIT_HAL_LOG_TIMESTAMP_MILLISECONDS,
+  MICROBIT_HAL_LOG_TIMESTAMP_MINUTES,
+  MICROBIT_HAL_LOG_TIMESTAMP_NONE,
+  MICROBIT_HAL_LOG_TIMESTAMP_SECONDS,
+} from "./constants";
+import { State } from "./state";
+
+// Determined via a CODAL program dumping logEnd - dataStart in MicroBitLog.cpp.
+// This is only approximate as we don't serialize our state in the same way but
+// it's important for the user to see that the log can fill up.
+const maxSizeBytes = 118780;
+
+export class DataLogging {
+  private mirroring: boolean = false;
+  private size: number = 0;
+  private timestamp = MICROBIT_HAL_LOG_TIMESTAMP_NONE;
+  private timestampOnLastEndRow: number | undefined;
+  private headingsChanged: boolean = false;
+  private headings: string[] = [];
+  private row: string[] | undefined;
+  private logFull: boolean = false;
+
+  constructor(
+    private currentTimeMillis: () => number,
+    private onLogOutput: (data: LogEntry) => void,
+    private onSerialOutput: (data: string) => void,
+    private onLogDelete: () => void,
+    private onChange: (state: Partial<State>) => void
+  ) {}
+
+  setMirroring(mirroring: boolean) {
+    this.mirroring = mirroring;
+  }
+
+  setTimestamp(timestamp: number) {
+    this.timestamp = timestamp;
+  }
+
+  beginRow() {
+    this.row = new Array(this.headings.length);
+    return MICROBIT_HAL_DEVICE_OK;
+  }
+
+  endRow() {
+    if (!this.row) {
+      throw noRowError();
+    }
+    if (this.timestamp !== this.timestampOnLastEndRow) {
+      // New timestamp column required. Put it first if there's been no output.
+      if (this.size === 0) {
+        this.headings = [timestampToHeading(this.timestamp), ...this.headings];
+        this.row = ["", ...this.row];
+      } else {
+        this.logData(timestampToHeading(this.timestamp), "");
```

Comment:
What's the purpose of this line?

Reply:
See https://github.com/lancaster-university/codal-microbit-v2/blob/master/source/MicroBitLog.cpp#L554